### ISSUE TITLE
Make C# threadpool compatible with Reactive pattern.

### ISSUE THREAD

**Issue by jtattermusch** [grpc/grpc#4880]
Currently, gRPC C# has a thread pool of 4 threads (hardcoded).

1. make threadpool size configurable
2. make sure that if long-running user-defined actions reacts to a grpc event, we eventually don't run out of threads that are waiting on grpc_completion_queue_next.

Probable solution:
Instead of having dedicated threads react to gRPC completion queue events, have a predefined number of Tasks wait for events. If an event is triggered by gRPC, make sure a new Task with completion_queue_next is spawned.

Before doing this, it would be good to collection some numbers on performance to be able verify that this approach doesn't hurt performance.

**Comment by jtattermusch** [comment]
#6498 exposes thread pool size setting as part of experimental API.

**Comment by manigandham** [comment]
You can use `Environment.ProcessorCount` for a good default size to start with.

https://msdn.microsoft.com/en-us/library/system.environment.processorcount(v=vs.110).aspx

**Comment by jtattermusch** [comment]
Yes, the change has been submitted and is part of 0.15.0 release. 
The current default is min(4, Environment.ProcessorCount / 2) which seemed to be a reasonable heuristic.
